### PR TITLE
cluster: support LockWithContext

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
This branches off https://github.com/mattermost/mattermost-plugin-api/pull/33 to add support for locking with a context (and thus implicitly supporting lock cancellation, "try lock" semantics, etc.).

I opted not to merge directly to `master` given the rework happening in that active PR.

### CODE REVIEW

File: cluster/mutex.go [hunk 1]
```diff
@@ -161,10 +162,23 @@ func (m *Mutex) refreshLock() error {
 // Lock locks m. If the mutex is already locked by any plugin instance, including the current one,
 // the calling goroutine blocks until the mutex can be locked.
 func (m *Mutex) Lock() {
+	m.LockWithContext(context.Background())
+}
+
+// Lock locks m unless the context is cancelled. If the mutex is already locked by any plugin
```
Comment (mickmister) [inline cluster/mutex.go]: Nit: function name in comment
```suggestion
// LockWithContext locks m unless the context is cancelled. If the mutex is already locked by any plugin
```

**Comment by lieut-data** [comment]
@ali-farooq0, is this something you can take a look at soon?

**Comment by ali-farooq0** [comment]
oh, the magical command doesn't work here ;/

**Pull request merged**